### ISSUE TITLE
Event Class Usability Issues

### ISSUE THREAD

**Issue by artemigkh** [meraki-analytics/cassiopeia#293]
https://github.com/meraki-analytics/cassiopeia/blob/master/cassiopeia/core/match.py#L405

1.  Type should probably be an enum
2.  The bigger problem: Each event returned from the API only has a small subset of the possible properties. The official DTO schema at https://developer.riotgames.com/api-methods/#match-v4/GET_getMatchTimeline has a lot of possibilities, but each event type only has 3-6 of these properties. Ex.
```
{
                    "type": "SKILL_LEVEL_UP",
                    "timestamp": 1201030,
                    "participantId": 8,
                    "skillSlot": 1,
                    "levelUpType": "NORMAL"
                },
                {
                    "type": "ITEM_DESTROYED",
                    "timestamp": 1203309,
                    "participantId": 6,
                    "itemId": 3513
                },
```

Accessing any of these (for example, `Event.tower_type` when event type is `SKILL_LEVEL_UP`) crashes program with an attribute exception, which I find hard to believe is intended behavior

If users are given a dict, then they can reasonably expect that with no schema the key/value pairs might be different each time, but with a framework like Cass that aims to make working with riot api data easy, it seems very un-intuitive that accessing a declared property on an object (the whole point is that Cass provides the schema so they don't have to spend time going through riot api documentation) causes their entire program to crash.

At the very least I feel like accessing properties not on the correct event type should return None instead of crashing, but ideally there should be support for common uses cases like currently exist with the participant timeline. Declare classes for common events such as `ChampionKillEvent`, `BuildingKillEvent`, `WardPlacedEvent`, etc. with python properties being only the ones that exist on that event type. Where possible, these classes should also contain corresponding Cass objects, such as lazy loaded `Summoner`s on  `ChampionKillEvent`. Then give the `Timeline` class functions like `ParticipantTimeline` except matchwide:
* `champion_kills-> List[ChampionKill]`
* `building_kills -> List[BuildingKillEvent]`
etc

I would be interested in implementing this set of functionality for match `TimeLine`, and I think it would be useful because Cass seems a bit difficult to use currently for people that want to analyze Match timelines. For example, if someone wanted a list of all the kills that happened in a game they would either have to concat all the events from all the frames, filter by `event.type = "CHAMPION_KILL"` (which they would have to look up somewhere because event types aren't available in a schema), then either look up somewhere such as the api or inherited `to_dict()` function to see what data points are actually available (and do a participantId lookup and mapping to get the summoner object if they wanted that). Or they could could go through the participant timelines, zip them by timestamp into a new list and filter duplicates, but either way functionality for analyzing match timelines as a whole should exist.

**Comment by jjmaldonis** [comment]
I'm going to close this as a "wontfix".

>Type should probably be an enum
It should be clear why that's not going to work by the end of this.

>The bigger problem: Each event returned from the API only has a small subset of the possible properties.
This is true regardless, and we aren't going to make this class a special case and allow inspection on it. You just have to know/figure out what events have what properties, and check for the event type. Riot doesn't document what fields get returned for what events, so everyone has to figure it out for themselves.

>Accessing any of these (for example, Event.tower_type when event type is SKILL_LEVEL_UP) crashes program with an attribute exception, which I find hard to believe is intended behavior
>At the very least I feel like accessing properties not on the correct event type should return None instead of crashing
We want Cass to crash if you access data that doesn't exist. Silently continuing the program will only introduce the bug further down the line. If you don't want your app to crash, then you use try/except blocks.

>ideally there should be support for common uses cases like currently exist with the participant timeline ... Then give the Timeline class functions like ParticipantTimeline except matchwide:
The participant timeline functionality is extremely complex, and integrating something like that into a datastructure as deep down as an Event is going to be worse. It might be possible, but I don't think it's worth the effort or the code complexity/maintainability.

>Declare classes for common events such as ChampionKillEvent, BuildingKillEvent, WardPlacedEvent, etc. with python properties being only the ones that exist on that event type.
This would likely be fine, but it comes at the cost of complexity/maintainability again. This is the only thing that makes sense if you don't want "nonexistent" properties on some events. However, you'll never know what events your getting when you're writing your code unless you check for a specific Event (e.g. isinstance(ChampionKillEvent); and if you're willing to do that, then just check `event.type` and only call the properties that exist for that `event.type`. Implementing classes might improve the documentation slightly, but it's at the cost of a bunch of new classes that we'd have to maintain perfectly.

>Cass seems a bit difficult to use currently for people that want to analyze Match timelines. For example ...
The timeline data is rough to use no matter what library you're using. All of the things you listed are going to need to be done regardless of the library. This is due to 1) the data Riot sends in the timeline and 2) how Riot structures that data.

In summary, while it would be nice for the Event class to be more user friendly, it is largely infeasible in my experience. We aren't going to break Cass's typesystem paradigms to implement something that seems slightly better on the surface, because it will only cause more confusion. Maybe subclassing Event into specific events could be useful, but it comes at the cost of significant maintainability issues that I don't want to take on right now.

**Comment by artemigkh** [comment]
I agree with everything you wrote, and indeed there are lots of problems with the things you addressed. 

However, I still think that it would be nice to have functions that make working with total match timeline easier. 

> The timeline data is rough to use no matter what library you're using. All of the things you listed are going to need to be done regardless of the library. This is due to 1) the data Riot sends in the timeline and 2) how Riot structures that data.

Yep, this is true, but Cassiopeia still does a lot of the heavy lifting when it comes to encapsulating participant timelines and making them easy and intuitive to work with, I don't see why there shouldn't be the same thing for match timelines as well. Not even doing anything overly complicated like in the original post, not adding any new classes or data structures, just providing functions exactly such as a list of all the events, a list of all the kills, etc that exist on `ParticipantTimeline`, but on the match timeline as well

**Comment by jjmaldonis** [comment]
Okay, that seems reasonable. As a side note, the `.filter` functionality in python/cass hopefully does most of the lifting for you.

I'm not remembering the timeline stuff very well, but doing something like:

`participant.cumulate_timeline["15:30"].events.filter(lambda event: event.type == 'CHAMPION_KILL')` seems like it should work/exist if it doesn't already.
